hack/generator/pkg/astmodel/armconversion/arm_conversion_function.go

```diff
@@ -0,0 +1,102 @@
+	"go/ast"
+)
+
+type ConversionDirection string
+
+const (
+	ConversionDirectionToArm   = ConversionDirection("ToArm")
+	ConversionDirectionFromArm = ConversionDirection("FromArm")
+)
+
+// ArmConversionFunction represents an ARM conversion function for converting between a Kubernetes resource
+// and an ARM resource.
+type ArmConversionFunction struct {
+	armTypeName astmodel.TypeName
+	armType     *astmodel.ObjectType
+	idFactory   astmodel.IdentifierFactory
+	direction   ConversionDirection
+	isResource  bool
+}
+
+var _ astmodel.Function = &ArmConversionFunction{}
+
+// RequiredImports returns the imports required for this conversion function
+func (c *ArmConversionFunction) RequiredImports() []astmodel.PackageReference {
+	var result []astmodel.PackageReference
+
+	// Because this interface is attached to an object, by definition that object will specify
+	// its own required imports. We don't want to call the objects required imports here
+	// because then we're in infinite recursion (object delegates to us, we delegate back to it)
+
+	// We need these because we're going to be constructing/casting to the types
+	// of the properties in the ARM object, so we need to import those.
+	result = append(result, c.armType.RequiredImports()...)
+	result = append(result, astmodel.MakeGenRuntimePackageReference())
+	result = append(result, astmodel.MakePackageReference("fmt"))
+
+	return result
+}
+
+// References this type has to the given type
+func (c *ArmConversionFunction) References() astmodel.TypeNameSet {
+	return c.armType.References()
+}
+
+// AsFunc returns the function as a Go AST
+func (c *ArmConversionFunction) AsFunc(codeGenerationContext *astmodel.CodeGenerationContext, receiver astmodel.TypeName, methodName string) *ast.FuncDecl {
+	switch c.direction {
+	case ConversionDirectionToArm:
+		return c.asConvertToArmFunc(codeGenerationContext, receiver, methodName)
+	case ConversionDirectionFromArm:
+		return c.asConvertFromArmFunc(codeGenerationContext, receiver, methodName)
+	default:
+		panic(fmt.Sprintf("Unknown conversion direction %s", c.direction))
+	}
+}
+
+func (c *ArmConversionFunction) asConvertToArmFunc(
+	codeGenerationContext *astmodel.CodeGenerationContext,
+	receiver astmodel.TypeName,
+	methodName string) *ast.FuncDecl {
+
+	builder := newConvertToArmFunctionBuilder(
+		c,
+		codeGenerationContext,
+		receiver,
+		methodName)
+	return builder.functionDeclaration()
+}
+
+func (c *ArmConversionFunction) asConvertFromArmFunc(
+	codeGenerationContext *astmodel.CodeGenerationContext,
+	receiver astmodel.TypeName,
+	methodName string) *ast.FuncDecl {
+
+	builder := newConvertFromArmFunctionBuilder(
+		c,
+		codeGenerationContext,
+		receiver,
+		methodName)
+	return builder.functionDeclaration()
+}
+
+// Equals determines if this function is equal to the passed in function
+func (c *ArmConversionFunction) Equals(other astmodel.Function) bool {
+	if o, ok := other.(*ArmConversionFunction); ok {
```

Comment:
Yeah, so equality for functions in general is actually a bit awkward.

1. Objects are immutable. So attaching a function to an object produces a new object. This makes passing a reference to the object into the function difficult (provided the function is _also_ immutable), as it's impossible to give two immutable objects references to each other. The current `OneOf` function actually has a "bug" here where the object it has a reference to is the object without itself attached.
2. The notion of attaching a function to an object (functions with receivers) sorta suggests that their equality only makes sense in the context of comparing objects - if this is the case then you end up in an infinite loop if the function tries to perform an equality check on the receiver objects, as that attempts to invoke the function Equals method again.

I'll add this in a note here, because it's definitely confusing. We can maybe think of an update to the object hierarchy (of functions) to change how things are represented to make this bits clearer.

Reply:
Leaving this comment open for further discussion although I'm not currently planning on doing anything else about it (other than added comment) in this PR.

Reply:
Ok, this makes sense, thanks!